legal-api/src/legal_api/services/filings/validations/alteration.py

```diff
@@ -0,0 +1,65 @@
+# Copyright © 2021 Province of British Columbia
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Validation for the Alteration filing."""
+from http import HTTPStatus
+from typing import Dict
+
+from flask_babel import _ as babel
+
+from legal_api.errors import Error
+from legal_api.models import Business, Filing
+
+from ... import namex
+from ...utils import get_str
+
+
+def validate(business: Business, filing: Dict) -> Error:
+    """Validate the Alteration filing."""
+    if not business or not filing:
+        return Error(HTTPStatus.BAD_REQUEST, [{'error': babel('A valid business and filing are required.')}])
+    msg = []
+
+    nr_path = '/filing/alteration/nameRequest/nrNumber'
+    nr_number = get_str(filing.json, nr_path)
+
+    if (nr_number) {
+        # ensure NR is approved or conditionally approved
+        nr_response = namex.query_nr_number(nr_number)
+        validation_result = namex.validate_nr(nr_response.json())
+        if not validation_result['is_approved']:
+            msg.append({'error': babel('Alteration of Name Request is not approved.'), 'path': nr_path})
+
+        # ensure business type is BCOMP
+        path = '/filing/alteration/nameRequest/legalType'
+        legal_type = get_str(filing, path)
+        if legal_type != Business.LegalTypes.BCOMP.value:
+            msg.append({'error': babel('Alteration of Name Request is not vaild for this type.'), 'path': path})
+
+        # ensure NR request has the same legal name
+        path = '/filing/alteration/nameRequest/legalName'
+        legal_name = get_str(filing, path)
+        nr_name = namex.get_approved_name(nr_response.json())
+        if nr_name != legal_name:
+            msg.append({'error': babel('Alteration of Name Request has a different legal name.'), 'path': path})
+    } else {
+        legal_name_path = '/filing/business/legalName'
+        legal_name = get_str(filing, legal_name_path)
+        if not legal_name:
+            msg.append({'error': babel('Alteration from Named to Numbered Company can only be done for a Named Company.'), 'path': legal_name_path})
+    }
```

Comment:
The test we did on Dev generated the following filing json when doing an alteration to a numbered company:

```
{
    "filing": {
        "header": {
            "date": "2021-02-11",
            "name": "alteration",
            "certifiedBy": ""
        },
        "business": {
            "legalName": "TESTING BENEFIT LTD.",
            "legalType": "BEN",
            "identifier": "BC1230085",
            "foundingDate": "2021-02-11T20:54:28.997767+00:00"
        },
        "alteration": {
            "business": {
                "legalType": "BEN",
                "identifier": "BC1230085"
            },
            "nameRequest": {
                "legalType": "BEN"
            },
            "contactPoint": {
                "email": "cameron@freshworks.io",
                "phone": "(123) 456-6555"
            },
            "shareStructure": {
                "shareClasses": [
                    {
                        "id": 431494,
                        "name": "Class A Shares",
                        "series": [],
                        "currency": "CAD",
                        "parValue": 1,
                        "priority": 1,
                        "hasParValue": true,
                        "hasMaximumShares": false,
                        "maxNumberOfShares": null,
                        "hasRightsOrRestrictions": false
                    }
                ],
                "resolutionDates": []
            },
            "nameTranslations": [],
            "provisionsRemoved": null
        }
    }
}
```
Notice we do have a nameRequest -> legalType 

Should we validate the legalType when we have nameRequest?

Reply:
AFAIK, alterations are specifically for LTD/ULC -> BEN changes.

It would be good to validate as much as possible:
- don't trust that the UI is providing valid data (*)
- protect yourself (db) from corrupted data

(*) At the moment it is only our UI posting data, but eventually the Legal API will be open to external UIs.

Reply:
Yes, I think you should validate that the NR is of the correct type.
(And I think that type is "BC" but please verify.)

Reply:
Is this resolved?

Reply:
According to the requirements there is no NR when going from named to numbered

Reply:
OK, I see this in the code now:
`# ensure NR request has the same legal name`

Looks like this is handled. Resolving this conversation.